lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -1139,39 +1140,87 @@ bool WitnessChecker::findBestWitness(ValueDecl *requirement,
                                      unsigned &numViable,
                                      unsigned &bestIdx,
                                      bool &doNotDiagnoseMatches) {
-  auto witnesses = lookupValueWitnesses(requirement, ignoringNames);
-
-  // Match each of the witnesses to the requirement.
-  bool anyFromUnconstrainedExtension = false;
-  numViable = 0;
-  bestIdx = 0;
-
-  for (auto witness : witnesses) {
-    // Don't match anything in a protocol.
-    // FIXME: When default implementations come along, we can try to match
-    // these when they're default implementations coming from another
-    // (unrelated) protocol.
-    if (isa<ProtocolDecl>(witness->getDeclContext())) {
-      continue;
-    }
+  enum Attempt {
```

Comment:
Ok, I'll stop complaining. Let's tack this onto @DougGregor's todo list for his upcoming name lookup cleanup. :)